### DISCUSSION TITLE
What kind of "process guidance" can we use for predicting DO?

### DISCUSSION THREAD

**Discussion by jsadler2** [USGS-R/drb-do-ml#15] [Ideas]
With the processes governing DO quite a bit more complex than other variables we've tried to predict, I think we will have to get creative with how we incorporate process guidance into the models.

**Comment by jsadler2** [comment]
# Correlation with temperature
This might be the most straightforward way to incorporate process understanding, since there is a known relationship between DO and water temperature.

**Reply by jsadler2** [reply]
One question I've had is if it even makes sense to impose a relationship between DO and temperature when the model will likely learn it anyway?

Maybe that's something we can test. Does the pure ML model actually learn the "correct" relationship between DO and temp on its own? In other words, I think it would be interesting to see if the relationship that is learned by a pure DL model lines up with the relationship that we know exists between DO and temp.

**Reply by galengorski** [reply]
This could be a way to assess performance. In addition to predictive performance, does the model "learn the theory" correctly, it could be a form of functional performance

**Comment by jzwart** [comment]
# Use `LakeMetabolizer` for DO estimates
Could use [LakeMetabolizer function](https://github.com/GLEON/LakeMetabolizer/blob/master/R/o2.at.sat.R) for calculating DO from temperature, pressure, and salinity. Returns expected DO in mg/L for 100% saturation (equilibrium). e.g.: 
> LakeMetabolizer::o2.at.sat.base(temp = 23, altitude = 515, salinity = 0)
[1] 8.054198

**Reply by jsadler2** [reply]
Ohh. I like this idea.  I do wonder how/where we could get reasonable salinity estimates. Any idea on that @jds485?

**Reply by lekoenig** [reply]
Thanks, @jzwart! I was also thinking about StreamMetabolizer's `calc_DO_sat` [function](https://github.com/USGS-R/streamMetabolizer/blob/master/R/calc_DO_sat.R), but it looks like `calc_DO_sat` calls the LakeMetabolizer function.

**Reply by lekoenig** [reply]
@jsadler2 salinity defaults to zero here so we could start with that assumption. Temperature will be a bigger driver of DO solubility in the range of salinity we observe in freshwater portions of the DRB.

**Reply by jds485** [reply]
This would be for pretraining data? Once a salinity prediction model is working, sounds like the outputs can be fed to here to generate this data. What timescale are you looking for? And what units (conductivity, concentration, other)? And I'd agree that starting with 0 is a good plan.

**Reply by jsadler2** [reply]
Yeah it could be for pretraining data or as an ~output~ input (https://github.com/USGS-R/river-dl/pull/142#issuecomment-963301429), @jds485 . 

And sounds like a good plan to start with salinity=0, @lekoenig

**Reply by jsadler2** [reply]
Questions I have about this (w/o knowing much about LakeMetabolizer or stream metabolism in the first place)
* does LakeMetabolizer implicitly account for the bio processes that affect DO, or does it ignore them? 
* I guess this simple function doesn't account for sunlight, which, from what I understand is a big part of GPP which is a big part of DO dynamics. Am I understanding that right?

**Reply by jzwart** [reply]
>- does LakeMetabolizer implicitly account for the bio processes that affect DO, or does it ignore them?
> - I guess this simple function doesn't account for sunlight, which, from what I understand is a big part of GPP which is a big part of DO dynamics. Am I understanding that right?

There are other functions in LakeMetabolizer / StreamMetabolizer that use sunlight and other data to estimate GPP and predicted DO. But the function above, `LakeMetabolizer::o2.at.sat()`, only calculates DO at equilibrium, which is then used for a bunch of other calculations in metabolism modeling (e.g. gas exchange, net ecosystem production).

**Reply by jzwart** [reply]
This would give the model a starting point if used as input / pretraining, but you'd need some other drivers to let the model predict actual DO such as sunlight, streamflow, wind speed,

**Comment by aappling-usgs** [comment]
Putting my vote in for persistence of metabolizing biomass over time (with disruptions especially during/after flood scour, and declines in the winter)

**Reply by jsadler2** [reply]
Sounds cool! .... Do you mind actually explaining that more, the "persistence of metabolizing biomass over time" part ? Is this the idea that one of the latent states of the model could start to behave like metabolizing biomass?

**Reply by lekoenig** [reply]
Thanks for sharing, Alison! I didn't really appreciate that we might be able to build latent states into the process guidance for these models. @jsadler2, perhaps we could chat more about that in one of our project meetings?

**Reply by aappling-usgs** [reply]
Yes, I think this would be a latent state thing. I'm not sure how to best implement this - can imagine tying one of the h variables in the LSTM to ~biomass via a loss term that looks right at that state and expects the DO amplitude to be within a range related to that state(where here the meaning would be "photosynthesizing biomass" rather that just "metabolizing"). Or could do something simpler about expecting (via loss) the change in DO amplitude from one day to the next to be limited (especially when it's on the upswing; biomass can crash during a scour event but rarely/never grows explosively). Curious to see what you, Lauren, and others can come up with.

**Reply by jsadler2** [reply]
Hm interesting ideas. To add a third, we could go the opposite direction of your first idea where we look at the predicted DO and use that to give a range to the h or c, adding to the loss if the latent state is outside the range.
